semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -43,36 +64,133 @@ firstLine :: Source -> Text
 firstLine = Text.takeWhile (/= '\n') . toText . Source.take 180
 
 
--- FIXME: move GFoldable1 into semantic-ast.
-class GFoldable1 c t where
-  -- | Generically map functions over fields of kind @* -> *@, monoidally combining the results.
-  gfoldMap1
-    :: Monoid b
-    => (forall f . c f => f a -> b)
+-- FIXME: move Traversable1 into semantic-ast.
+-- FIXME: derive Traversable1 instances for TH-generated syntax types.
+
+-- | Simultaneous traversal of subterms of kind @*@ and @* -> *@ in an 'Applicative' context.
+--
+-- 'Traversable1' can express any combination of first- and second-order mapping, folding, and traversal.
+class Traversable1 c t where
+  -- | Map annotations of kind @*@ and heterogeneously-typed subterms of kind @* -> *@ under some constraint @c@ into an 'Applicative' context. The constraint is necessary to operate on otherwise universally-quantified subterms, since otherwise there would be insufficient information to inspect them at all.
+  --
+  -- No proxy is provided for the constraint @c@; instead, @-XTypeApplications@ should be used. E.g. here we ignore the annotations and print all the @* -> *@ subterms using 'Show1':
+  --
+  -- @
+  -- 'traverse1' \@'Data.Functor.Classes.Show1' 'pure' (\ t -> t '<$' 'putStrLn' ('Data.Functor.Classes.showsPrec1' 0 t ""))
+  -- @
+  --
+  -- Note that this traversal is non-recursive: any recursion through subterms must be performed by the second function argument.
+  traverse1
+    :: Applicative f
+    => (a -> f b)
+    -> (forall t' . c t' => t' a -> f (t' b))
+    -> t a
+    -> f (t b)
+  default traverse1
+    :: (Applicative f, Generic1 t, GTraversable1 c (Rep1 t))
+    => (a -> f b)
+    -> (forall t' . c t' => t' a -> f (t' b))
     -> t a
-    -> b
+    -> f (t b)
+  traverse1 f g = fmap to1 . gtraverse1 @c f g . from1
+
+for1
+  :: forall c t f a b
+  .  (Traversable1 c t, Applicative f)
+  => t a
+  -> (a -> f b)
+  -> (forall t' . c t' => t' a -> f (t' b))
+  -> f (t b)
+for1 t f g = traverse1 @c f g t
+
+traverse1_
+  :: forall c t f a a' a''
+  .  (Traversable1 c t, Applicative f)
+  => (a -> f a')
+  -> (forall t' . c t' => t' a -> f a'')
+  -> t a
+  -> f ()
+traverse1_ f g = getAp . foldMap1 @c (Ap . void . f) (Ap . void . g)
+
+for1_
+  :: forall c t f a a' a''
+  .  (Traversable1 c t, Applicative f)
+  => t a
+  -> (a -> f a')
+  -> (forall t' . c t' => t' a -> f a'')
+  -> f ()
+for1_ t f g = getAp $ foldMap1 @c (Ap . void . f) (Ap . void . g) t
+
+foldMap1 :: forall c t b a . (Traversable1 c t, Monoid b) => (a -> b) -> (forall t' . c t' => t' a -> b) -> t a -> b
+foldMap1 f g = getConst . traverse1 @c (Const . f) (Const . g)
```

Comment:
I’ve removed `GFoldable1` instead of defining `Foldable1` and making it a superclass of `Traversable1` largely because we are focusing our attention on the ASTs we get out of TH. We might eventually decide that this is useful enough to warrant some more interesting hierarchy, but I kind of doubt that there are many cases where we’d really have non-`Traversable1` `Foldable1` types.